design-documents/database/declarative-schema-database-compatibility.md

```diff
@@ -0,0 +1,207 @@
+# Declarative schema database compatibility
+
+## About
+
+Declarative schema relies on querying metadata from the DB.
+The query itself and format of the result may differ for different RDBMS systems.
+For example, here is the query from `Magento\Framework\Setup\Declaration\Schema\Db\MySQL\DbSchemaReader::readColumns`
+
+```
+SELECT 
+    COLUMN_NAME as 'name',
+    COLUMN_DEFAULT as 'default',
+    DATA_TYPE as 'type',
+    IS_NULLABLE as 'nullable',
+    COLUMN_TYPE as 'definition',
+    EXTRA as 'extra',
+    COLUMN_COMMENT as 'comment'
+FROM
+    information_schema.COLUMNS
+WHERE
+    TABLE_SCHEMA = 'magento'
+AND
+    TABLE_NAME = 'store_website'
+ORDER BY
+    ORDINAL_POSITION ASC;
+```
+
+For `COLUMN_DEFAULT`, PHP PDO connector returns `NULL` for MySQL and `"NULL"` (string) for MariaDB 10.2.7+.
+
+## Design
+
+Introduce new interfaces under Db namespace that would allow to get information about table schema, refactor implementations of `\Magento\Framework\Setup\Declaration\Schema\Db\DbSchemaReaderInterface` in declarative schema to use interfaces under Db namespace.
+
+Revise implementation of these interfaces to see if anything need to be changed here
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DDLTriggerInterface`
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DbDefinitionProcessorInterface` (see usages of `\Magento\Framework\Setup\Declaration\Schema\Db\DefinitionAggregator`)
+
+New interfaces and types for managing table schema.
+
+```
+class InformationSchema\Table
+{
+    public function getSchema(): string {}
+    public function getName(): string {}
+    public function getEngine(): string {}
```

Comment:
Propose to either to remove it or introduce enum for engine. I think we need to introduce enums for collation and charset. What is schema, definition of the table?

Reply:
I would not use an enum for the engine. Technically most RDBMS support having storage engines other than their default but this is mostly used in MySQL / MariaDB. Introducing an enum for engines could become a maintenance issue as MySQL / MariaDB evolve (e.g. as of MariaDB 10.4 the default storage engine is AriaDB which is an evolution of MyISAM).

Schema is (more or less) the database in which the table resides. SQL defines that a table resides in the schema of a catalog, hence the columns `information_schema.tables.catalog` and `information_schema.tables.schema`. MySQL / MariaDB do not support catalogs thus the schema reflects the database (as opposed to PostgreSQL supporting multiple schemas per catalog).